contract_comm/currency/currency.go

```diff
@@ -94,97 +94,127 @@ const (
+	rate, err := NewExchangeRate(numerator, denominator)
+	if err != nil {
+		panic(err)
 	}
+	return rate
+}
 
-	celoGoldAddress, err := contract_comm.GetRegisteredAddress(params.GoldTokenRegistryId, nil, nil)
-	if err == errors.ErrSmartContractNotDeployed || err == errors.ErrRegistryContractNotDeployed {
-		log.Warn("Registry address lookup failed", "err", err)
-		return val, err
+// NewExchangeRate creates an exchange rate.
+// Requires numerator >=0 && denominator >= 0
+func NewExchangeRate(numerator *big.Int, denominator *big.Int) (*ExchangeRate, error) {
+	if numerator == nil || common.Big0.Cmp(numerator) >= 0 {
+		return nil, errors.ErrExchangeRateZero
 	}
+	if denominator == nil || common.Big0.Cmp(denominator) >= 0 {
+		return nil, errors.ErrExchangeRateZero
+	}
+	return &ExchangeRate{numerator, denominator}, nil
+}
 
-	if *currencyFrom == celoGoldAddress {
-		// This function shouldn't really be called with the token's address, but if it is the value
-		// is correct, so return nil as the error
-		return val, nil
-	} else if err != nil {
-		log.Error(err.Error())
-		return val, err
+// ToBase converts from token to base
+func (er *ExchangeRate) ToBase(tokenAmount *big.Int) *big.Int {
+	// check noop rate (cGLD rate)
+	if er == nil {
+		return new(big.Int).Set(tokenAmount)
 	}
-	return Convert(val, currencyFrom, nil)
+
+	return new(big.Int).Div(new(big.Int).Mul(tokenAmount, er.denominator), er.numerator)
 }
 
-// NOTE (jarmg 4/24/19): values are rounded down which can cause
-// an estimate to be off by 1 (at most)
-func Convert(val *big.Int, currencyFrom *common.Address, currencyTo *common.Address) (*big.Int, error) {
-	exchangeRateFrom, err1 := getExchangeRate(currencyFrom)
-	exchangeRateTo, err2 := getExchangeRate(currencyTo)
+// FromGold converts from base to token
+func (er *ExchangeRate) FromBase(goldAmount *big.Int) *big.Int {
+	// check noop rate (cGLD rate)
+	if er == nil {
+		return new(big.Int).Set(goldAmount)
+	}
 
-	if err1 != nil || err2 != nil {
-		log.Error("Convert - Error in retreiving currency exchange rates")
-		if err1 != nil {
-			return nil, err1
-		}
-		if err2 != nil {
-			return nil, err2
-		}
+	return new(big.Int).Div(new(big.Int).Mul(goldAmount, er.numerator), er.denominator)
+}
+
+func (er *ExchangeRate) CmpValues(amount *big.Int, anotherTokenAmount *big.Int, anotherTokenRate *ExchangeRate) int {
+	// if both rates are noop rate (cGLD rate), compare values
+	if er == nil && anotherTokenRate == nil {
+		return amount.Cmp(anotherTokenAmount)
 	}
 
-	// Given value of val and rates n1/d1 and n2/d2 the function below does
-	// (val * d1 * n2) / (n1 * d2)
-	numerator := new(big.Int).Mul(val, new(big.Int).Mul(exchangeRateFrom.Denominator, exchangeRateTo.Numerator))
-	denominator := new(big.Int).Mul(exchangeRateFrom.Numerator, exchangeRateTo.Denominator)
+	// if both exchangeRate are the same
+	if er != nil && anotherTokenRate != nil && *er == *anotherTokenRate {
```

Comment:
I don't think this short-circuit works as intended, because the values in `ExchangeRate` are pointers (`*big.Int`).  We could add a `.Cmp()`.  But I also don't really like the signature of this function.  I would prefer for them to be a type that indicates an `amount,exchangeRate` pair, and then that type would have a `.Cmp()` function.

Reply:
On some intermediate version of this, i did had `TokenValue{amount, rate}`; since as you said, seems cleaner.

But the caller of this function is the txpool, so, who starts with (amount, feeCurrencyAddress); so it require the txpool to first fetch the exchangeRate, then create a TokenValue, then do the same for the other tx you want to compare with, and finally call `tokenValue.Cmp(tokenValue)`.

In the err, i've err on making txpool life simpler.

Another thing i've tried is leaving the signature as it is, but implementing it with `TokenValue`; but seemed that i was just creating the type to make this comparison, and i was adding complexity just becuase, so i removed the whole thing.

Reply:
On the comparison, you're right, but it's intented. I was trying to short circuit the case where they are actually the same instance.. and if not i do the normal comparison.

I though that if needed to do the actual Numerator and Denominator comparison; then i was not winning much. And thus i should just do the normal operation. 

But that was just an intuition, wdyt?

Reply:
Makes sense.  As far as when and how much to optimize, I don't know.  But if we want to short-circuit if the instances are the same, I think we could do a single check for `er == anotherTokenRate`.  I think this would be ok both as is and if we replace the `nil` rate with a {1, 1} rate.